indexer/src/main.ts

```diff
@@ -153,37 +173,104 @@ export default async function transform({
-      const receiptBloom = new Bloom(hexToBytes(ethReceipt.logsBloom));
-      blockLogsBloom.or(receiptBloom);
       cumulativeGasUsed += BigInt(ethReceipt.gasUsed);
+      // ethTx.transactionIndex can be null (if the block is pending) but
+      // Number(null) is 0 so this won't panic.
+      cumulativeGasUsages[Number(ethTx.transactionIndex)] = cumulativeGasUsed;
 
       // Add all the eth data to the store.
       store.push({ collection: "transactions", data: { tx: ethTx } });
       store.push({ collection: "receipts", data: { receipt: ethReceipt } });
       ethLogs.forEach((ethLog) => {
         store.push({ collection: "logs", data: { log: ethLog } });
       });
-    }),
+
+      // Add the logs bloom of the receipt to the block logs bloom.
+      const receiptLogsBloom = new Bloom(hexToBytes(ethReceipt.logsBloom));
+      blockLogsBloom.or(receiptLogsBloom);
+
+      /// Return the trie data.
+      return createTrieData({
+        transactionIndex: Number(ethTx.transactionIndex),
+        typedTransaction: typedEthTx,
+        receipt: ethReceipt,
+      });
+    },
+  );
+
+  // Filter out the null values for the trie data.
+  const trieData = maybeTrieData.filter((x) => x !== null) as Array<TrieData>;
+
+  // Compute the blooms in an async manner.
+  await Promise.all(
+    trieData.map(
+      async ({
+        encodedTransactionIndex,
+        encodedTransaction,
+        encodedReceipt,
+      }) => {
+        // Add the transaction to the transaction trie.
+        await transactionTrie.put(encodedTransactionIndex, encodedTransaction);
+        // Add the receipt to the receipt trie.
+        await receiptTrie.put(encodedTransactionIndex, encodedReceipt);
+      },
+    ),
   );
 
+  // Sort the cumulative gas uses by descending transaction index.
+  cumulativeGasUsages.reverse();
+
+  (transactions ?? []).forEach(({ transaction, receipt }) => {
+    if (isRevertedWithOutOfResources(receipt)) {
+      // Can be false if the transaction is not related to a specific instance of the Kakarot contract.
+      // This is typically the case if there are multiple Kakarot contracts on the same chain.
+      const isKakarotTx = isKakarotTransaction(transaction);
+      if (!isKakarotTx) {
+        return;
+      }
+
+      const ethTx = toEthTx({
+        transaction,
+        receipt,
+        blockNumber,
+        blockHash,
+        isPendingBlock,
+      });
+      if (!ethTx) {
+        return;
+      }
+
+      // Get the cumulative gas used for the reverted transaction.
+      // Example:
+      // const cumulativeGasUsages = [300n, undefined, undefined, 200n, undefined, 100n, undefined, undefined, 10n, undefined];
+      // const ethTx = { transactionIndex: 5 };
+      // const revertedTransactionCumulativeGasUsed = 100n;
+      const len = cumulativeGasUsages.length;
+      const revertedTransactionCumulativeGasUsed =
+        cumulativeGasUsages.find((gas, i) => {
```

Comment:
how is it possible to have undefined at the tx index?

Reply:
If the transaction was reverted due to run out of resources error no? But we want to index these failures in order to provide user feedback.

Reply:
the way we fill `cumulativeGasUsages`, we set `cumulativeGasUsages[Number(ethTx.transactionIndex)] = cumulativeGasUsed;` However, `ethTx.transactionIndex` is not necessarily sequential: because it's based on Starknet transactions, and you could have some Starknet transactions reverting (with out of resources for example), the `ethTx.transactionIndex` could be 1, 5, 10, skipping some values. This will cause the `cumulativeGasUsages` to look something like this: `[undefined, 10n, undefined, undefined, 50n, undefined, 100n, ...]`. Thus we want to find the first value in this array that isn't undefined and has a lower index than the reverted Starknet tx.

This is also why I was using `{index: number, cumulativeGas: bigint}` as I thought this made the goal of this array clearer.

Reply:
I didn't understand first that for the reverted tx at stake there is no gas used, so that we need to pick the next one well defined.
I think that with a more functional approach this can be made simpler, so let's keep this for now and refacto during #1187